src/crimson/osd/heartbeat.cc

```diff
@@ -304,39 +304,38 @@ seastar::future<> Heartbeat::maybe_share_osdmap(
 {
   const osd_id_t from = m->get_source().num();
   const epoch_t osdmap_epoch = service.get_map()->get_epoch();
-  const epoch_t peer_epoch = m->map_epoch;
   auto found = peers.find(from);
   if (found == peers.end()) {
     return seastar::now();
   }
   auto& peer = found->second;
 
-  if (peer_epoch > peer.get_last_epoch_sent()) {
+  if (m->map_epoch > peer.get_last_epoch_sent()) {
     logger().debug("{} updating session's last epoch sent "
                    "from {} to peer's (id: {}) map epoch of {}",
                    __func__, peer.get_last_epoch_sent(),
-                   from, peer_epoch);
-    peer.set_last_epoch_sent(peer_epoch);
+                   from, m->map_epoch);
+    peer.set_last_epoch_sent(m->map_epoch);
```

Comment:
What I mean is that I think we can call

```
      peer.set_last_epoch_sent(service.get_map()->get_epoch());
```

here.  That would prevent subsequent calls to maybe_share_osdmap from sending duplicate maps.  The portion after send_incremental_map_to_osd would be unnecessary.

Reply:
The first `set_last_epoch_sent` is there to prevent us from sending already sent epochs. Adding @liu-chunmei's example below:

`m->map_epoch=4`, `peer.last_epoch_sent=3`, `last_epoch_sent` will be updated to 4.
`osdmap_epoch=6`, we will send_incremental_map (4 to 6) - and `last_epoch_sent` will be updated to **6**.

_Next ping_:
`m->map_epoch=4` and `osdmap_epoch=8`, we will send_incremental_map (6 to 8) **instead** of (4 to 8).

---

If we will set_last_epoch_sent to `service.get_map()->get_epoch()` here, we wouldn't be able to identify whether we need to `send_incremental_map_to_osd` or not.

Only after identifying that, we update last_epoch_sent to `service.get_map()`.

Reply:
By "If we will set_last_epoch_sent to ```service.get_map()->get_epoch()``` here, we wouldn't be able to identify whether we need to send_incremental_map_to_osd or not.", do you mean because we then invoke ```service.send_incremental_map_to_osd(from, peer.get_last_epoch_sent())```?

It seems like it should be fine so long as change that call to use the previous last_epoch_sent.  I put together a branch (https://github.com/athanatos/ceph/tree/sjust/wip-crimson-osdmap-heartbeat-61304) with that change and a couple of other cleanups.  Does that make sense, or am I missing something else?

Reply:
> By "If we will set_last_epoch_sent to `service.get_map()->get_epoch()` here, we wouldn't be able to identify whether we need to send_incremental_map_to_osd or not.", do you mean because we then invoke `service.send_incremental_map_to_osd(from, peer.get_last_epoch_sent())`?

IIUC, your suggestion is to call:
```
peer.set_projected_epoch(current_osdmap_epoch);
```
However, the following if-statement will always be true (hence _"..we wouldn't be able to identify whether we need to send or not"_)

```
if (current_osdmap_epoch <= peer.get_projected_epoch()) {
  return seastar::now();
}
```


> It seems like it should be fine so long as change that call to use the previous last_epoch_sent. I put together a branch (https://github.com/athanatos/ceph/tree/sjust/wip-crimson-osdmap-heartbeat-61304) with that change and a couple of other cleanups. Does that make sense, or am I missing something else?

I will combine both branches, thank you!

Reply:
My version only calls ```peer.set_projected_epoch(current_osdmap_epoch);``` after that check.

Reply:
@athanatos Got it now.